### PR TITLE
allowed updates

### PR DESCRIPTION
**Pull request by WaffleLapkin**
- Make handler a named struct (instead of named tuple)
- Add ability to automatically infer interesting updates

### CODE REVIEW

**Comment by WaffleLapkin** [comment]
To be clear: my current naming is bad. Maybe you have better ideas?

File: examples/descr_locations.rs [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+    #[track_caller]
+    fn filter_map_async() -> Self {
+        collector("filter_map_async")
+    }
+
+    #[track_caller]
+    fn endpoint() -> Self {
+        collector("endpoint")
+    }
```
Comment (Hirrolot) [inline examples/descr_locations.rs:84]: Maybe generate these definitions via a macro? Something like `methods! { map, map_async, filter, filter_async, ... };` ?

File: src/handler/description.rs [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+    /// [`user_defined`](HandlerDescription::user_defined).
+    #[track_caller]
+    fn endpoint() -> Self {
+        Self::user_defined()
+    }
+}
+
+/// Uninformative handler description.
+#[derive(Debug, PartialEq, Eq)]
+pub struct Unspecified(());
```
Comment (Hirrolot) [inline src/handler/description.rs:151]: `pub struct Unspecified;` ?

File: src/handler/description.rs [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+
+/// Description for a handler that describes what event kinds are interesting to
+/// the handler.
+#[derive(Debug, Clone)]
+pub enum EventKindDescription<K, S = RandomState> {
+    /// Only event) kinds in the set are "interesting".
```
Comment (Hirrolot) [inline src/handler/description.rs:157]: A doc typo?

File: src/handler/description.rs [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+    InterestingEventKinds(HashSet<K, S>),
```
Comment (Hirrolot) [inline src/handler/description.rs:158]: "Interesting" sounds a bit peculiar to me. But maybe it's just me. It seems that `epoll` uses the term "interest list", which is almost the same as we use here.

**Review by Hirrolot — CHANGES_REQUESTED** [review]
In general, a very interesting API and a very powerful idea, congratulations!

Just think of `dptree` overall:

 - It implements the chain-of-responsibility pattern that scales to tree-of-responsibility is needed;
 - It provides a poor man's DI;
 - It allows to investigate its own structure with ease.

Looks like a foundation of a brand new programming language or something like this.

File: examples/descr_locations.rs (hunk 1 above)
Comment (WaffleLapkin) [reply examples/descr_locations.rs:84]: I don't think this is needed in this case, macro will only obscure the code

File: src/handler/description.rs (hunk 2 above)
Comment (WaffleLapkin) [reply src/handler/description.rs:151]: `(())` is the standard way of making the struct not constractable (it is a private field).

File: src/handler/description.rs (hunk 4 above)
Comment (WaffleLapkin) [reply src/handler/description.rs:158]: I'm not a fun of "interesting" either, but I don't know how to better describe "possibly observable by the endpoints or user defined handlers (`from_fn`)" 😅

File: src/handler/description.rs (hunk 3 above)
Comment (WaffleLapkin) [reply src/handler/description.rs]: Yep, fixed.

File: src/handler/description.rs [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+/// Uninformative handler description.
+#[derive(Debug, PartialEq, Eq)]
+pub struct Unspecified(());
+
+/// Description for a handler that describes what event kinds are interesting to
+/// the handler.
+#[derive(Debug, Clone)]
+pub enum EventKindDescription<K, S = RandomState> {
+    /// Only event kinds in the set are "interesting".
+    InterestingEventKinds(HashSet<K, S>),
```
Comment (Hirrolot) [inline src/handler/description.rs:158]: Still I suggest changing `InterestingEventKinds` => `InterestList`. As described by `epoll`:

```
The interest list (sometimes also called the epoll set): the set
of file descriptors that the process has registered an interest in monitoring.
```

It seems that it is a more concise term.

**Review by Hirrolot — CHANGES_REQUESTED** [review]

**Review by Hirrolot — APPROVED** [review]

File: src/handler/description.rs [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+    /// Description for [`endpoint`](crate::endpoint).
+    ///
+    /// ## Default implementation
+    ///
+    /// By default this returns the value from
+    /// [`user_defined`](HandlerDescription::user_defined).
+    #[track_caller]
+    fn endpoint() -> Self {
+        Self::user_defined()
+    }
```
Comment (Hirrolot) [inline src/handler/description.rs:146]: Is there any place where these functions are needed except the example `descr_locations.rs`?

Comment (WaffleLapkin) [reply src/handler/description.rs:146]: Not currently, no. But I do believe that `descr_locations.rs` is a good example. You could imagine a similar thing that literally draws the tree and allow you to click through it into source code.

**Pull request merged**